tensionmap.py

```diff
@@ -147,35 +154,60 @@ def tm_update(obj, context):
-
+    
+    #create vertex color list for faster access
+    number_of_tm_channels = 2
+    vertex_colors = [0]*(number_of_tm_channels*num_vertices)
     # put the new values in the vertex groups
     for i in range(num_vertices):
         add_index = [i]
+        stretch_value = obj.data.tm_minimum
+        squeeze_value = obj.data.tm_minimum
         if weights[i] >= 0:
             # positive: stretched
-            group_squeeze.add(add_index, obj.data.tm_minimum, "REPLACE")
-            group_stretch.add(add_index, max(obj.data.tm_minimum, min(obj.data.tm_maximum, weights[i])), "REPLACE")
+            stretch_value = max(obj.data.tm_minimum, min(obj.data.tm_maximum, weights[i]))
+            if obj.data.tm_enable_vertex_groups:
+                group_squeeze.add(add_index, obj.data.tm_minimum, "REPLACE")
+                group_stretch.add(add_index, stretch_value, "REPLACE")
         else:
             # negative: squeezed
             # invert weights to keep only positive values
-            group_squeeze.add(add_index, max(obj.data.tm_minimum, min(obj.data.tm_maximum, -weights[i])), "REPLACE")
-            group_stretch.add(add_index, obj.data.tm_minimum, "REPLACE")
+            squeeze_value = max(obj.data.tm_minimum, min(obj.data.tm_maximum, -weights[i]))
+            if obj.data.tm_enable_vertex_groups:
+                group_squeeze.add(add_index, squeeze_value, "REPLACE")
+                group_stretch.add(add_index, obj.data.tm_minimum, "REPLACE")
+        vertex_colors[i*number_of_tm_channels] = stretch_value  # red
+        vertex_colors[i*number_of_tm_channels+1] = squeeze_value  # green
 
     if obj.data.tm_enable_vertex_colors:
         colors_tension = get_or_create_vertex_colors(obj, "tm_tension")
         # put the new values from the vertex groups in the vertex colors
         # this is heavy, but vertex colors are stored by vertex loop
         # and there is no simpler way to do it (it would seem)
-        for i in range(len(obj.data.polygons)):
-            polygon = obj.data.polygons[i]
-            for key, value in enumerate(polygon.loop_indices):
-                vertex_color = colors_tension.data[value]
-                vertex = polygon.vertices[key]
-
-                vertex_color.color[0] = group_stretch.weight(vertex)  # red
-                vertex_color.color[1] = group_squeeze.weight(vertex)  # green
-                vertex_color.color[2] = 0.0  # blue
-
+        for poly_idx in range(len(obj.data.polygons)):
+            polygon = obj.data.polygons[poly_idx]
+            for loop_vertex_idx, loop_idx in enumerate(polygon.loop_indices):
+                vertex_color = colors_tension.data[loop_idx]
+                vertex_idx = polygon.vertices[loop_vertex_idx]
+                vertex_color.color[0] = vertex_colors[vertex_idx*number_of_tm_channels]
+                vertex_color.color[1] = vertex_colors[vertex_idx*number_of_tm_channels+1]
+
+                        
+def tm_update_original_edges(obj):
```

Comment:
I see a few problems with this function.

First, I do not see how you can improve performance with it. This function is executed as many times as the `tm_update` function, which used to handle this logic.

Plus, you are now introducing a second loop. The previous implementation was using a single loop to compute the original & the deformed edge length. So I see it as a performance decrease.

And finally, using a global variable for this is ugly.
If you wanted to do it nicely, this function would return the `original_edge_lengths`, and `tm_update` would then take it as a parameter.

Reply:
The performance increase is due to the function being called only by `tm_update_selected`, which is only called when clicking the "Update Tension Map" button. It is not called by `tm_update` each frame.

The global variable is very ugly and also the source of another problem: initialization.
The script keeps throwing `IndexError: list index out of range` until you press the button, because `original_edge_lengths` starts empty!
That's what I get for lazily using a global variable after decades of rightful abstinence! 😃

I'll see how I fix that.

Reply:
Oh I see.
But while that might increase performance, it comes at the great cost of usability. It means you will need to manually update your meshes when you make changes to them.

Also, since I saw 0% of performance increase when using the Vertex Group, I would suppose that this does not come with any performance improvement.
Otherwise, the Vertex Group generation would also be faster, not just the Vertex Color.

What do you think ?

Reply:
Hm, I see. Having to click that button is not a viable solution, of course.
I suppose there must be a way to detect if an object mesh has changed and I already found some [first possibilities](https://docs.blender.org/api/blender2.8/bpy.types.DepsgraphUpdate.html?highlight=is_updated#bpy.types.DepsgraphUpdate.is_updated_geometry).
I'll explore that.

Disabling the Vertex Group generation brought a little bit of performance for me, but it was only a  increase from 15.5 to 16.5 fps on a model with 2000 vertices. The main advantage in my eyes is that you don't need to have two additional vertex groups "polluting" your list if you don't need them.

Reply:
The depsgraph was a dead end, it seems to be related to the modifier stack. Apparently I would have to hash the base model/vertex data in some way to notice it changed and I guess this would be too costly.

I changed the implementation of the edge calculation so that it still is a global variable, but done via a dictionary, containing each object for which the Tension Map is enabled.
It checks during the update if it has seen the object. If not, it generates the data once.
If the Tension Map is disabled, the edge list gets deleted from the dictionary.
If the geometry of an object changed, the user has to click "Update Tension Map" button.
I guess that is the best solution for now. If you have better suggestions for both, I'm all ears. 😃

Reply:
> Disabling the Vertex Group generation brought a little bit of performance for me, but it was only a increase from 15.5 to 16.5 fps on a model with 2000 vertices. The main advantage in my eyes is that you don't need to have two additional vertex groups "polluting" your list if you don't need them.

`Vertex Group + Vertex Color` VS `only Vertex Color` definitively had performance improvements, as I said in my post about performance.
What I am talking about is `the old code with only Vertex Group` VS `the new code with only Vertex Group`.

In that case, I saw no difference in performance.
What that tells me is, the optimization we are talking about in this thread doesn't seem to actually improve performance, otherwise, the new code would be faster than the old one, even if it's only generating the Vertex Group.

Reply:
Strange, the optimization with the global var is one of the main performance improvements!

And, I mean, it makes sense:
If vertex group or color generation is enabled, we need to calculate stretch/squeeze during each frame in both cases.
What the global variable stuff does, is moving the calculation of the original edge lengths out of the update function, so we don't have to do that each frame. It is only evaluated if the user presses the button or if we need it the first time.

On my system, simply replacing line 149
`original_edge_length = original_edge_lengths[i]`
with 
`original_edge_length = original_edge_lengths[i] = (obj.data.vertices[first_vertex].co - 
                                    obj.data.vertices[second_vertex].co).length`
(thus bypassing the improvements and calculating it locally again) raises my fps from 30 to 34, with "Enable Vertex Groups" only.
Please try it on your system, maybe some other changes were influencing the result.

Also you are right, I should make one commit per improvement, that makes testing easier, too.

Reply:
I'm back. I've made [an addon](https://github.com/ScottishCyclops/viewport-fps) to accurately measure the average FPS.

Average FPS over 600 frames, calculated in my test scene for each versions:

Options | 2 1 1 | 2 1 1 no global | 2 1 0
-- | -- | -- | --
All | 11,37 | 11,09 | 8,87
Group only | 12,79 | 12,4 | 12,67
Color only | 11,68 | 11,2 |  

Comparisons (x vs y, x is how much faster in percent)

2 1 1 vs 2 1 1 no global | 2 1 1 vs 2 1 0 | 2 1 1 no global vs 2 1 0
-- | -- | --
2,52 % | 28,18 % | 25,03 %
3,15 % | 0,95 % | -2,13 %
4,29 % |   |  


> Strange, the optimization with the global var is one of the main performance improvements!

So the global optimization (let's refer to it as the "global optimization" from now on) is definitively not the main performance improvement. Without it, we still get a beautiful >25% overall performance increase.

That said, it is still a bit faster, so it is a good idea to explore.

I also noted that with the new code and without the *global optimization*, only calculating the Vertex Group is a tiny bit slower. What that means is your code is better at calculating the Vertex Color, but for Vertex Group, is adds overhead.

Can't help it, since we create additional arrays and such.

Here is what I propose: we ignore the  *global optimization* for this PR, and merge the rest of the improvements.

Then, you can continue working on a good solution to integrate the *global optimization*, and make a new PR for it.

In the meantime, I saw other things that might improve the perf a little for Vertex Group, so I'll work on that.

Reply:
Okay, I ran it, too and I got similar results:

|            | 211  | 211 no Global | 210  |
|------------|------|---------------|------|
| All        | 20,8 | 19,5          | 11,5 |
| Group Only | 31,6 | 28,6          | 30,3 |
| Color Only | 21,9 | 21,1          |      |

You're right, the influence of the global optimization variant has a smaller influence than I thought.
I will commit the 211 no Global variant I created for that test, so you can simply merge that.

Reply:
>  PS: There is a small typo, the toolbar says "Wiewport".

On my way to fix that x)